### PR TITLE
Replace Lucene DataInput/DataOutput with Elasticsearch StreamInput/StreamOutput

### PR DESCRIPTION
**Pull request by romseygeek**
In a number of places, we read and write binary data into byte arrays using lucene's 
DataInput and DataOutput abstractions. In lucene 9 these abstractions are changing
the endianness of their read/writeInt methods. To avoid dealing with this formatting
change, this commit changes things to use elasticsearch StreamInput/StreamOutput
abstractions instead, which have basically the same API but will preserve endianness.

Relates to #73324

### CODE REVIEW

**Comment by romseygeek** [comment]
We still use DataInput/DataOutput legitimately in a few places, particularly around the translog, so I decided against adding the lucene classes to forbidden APIs.

File: server/src/main/java/org/elasticsearch/common/util/ByteUtils.java [hunk 1]
```diff
@@ -85,45 +82,4 @@ public static float readFloatLE(byte[] arr, int offset) {
         return Float.intBitsToFloat(readIntLE(arr, offset));
     }
 
-    /** Same as DataOutput#writeVLong but accepts negative values (written on 9 bytes). */
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/common/util/ByteUtils.java:88]: These two methods were unused everywhere except for tests, so I removed them and the tests as well.

**Comment by romseygeek** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Review by iverase — APPROVED** [review]
LGTM, just add a suggestion about throwing an UncheckedIOException.

Note to us: We did not address the following usage which I think it is the right thing to do in this PR but it might break when we update to lucene 9. Checksums are always stored as Big Endian.

https://github.com/elastic/elasticsearch/blob/6d99735ab82f185a45c5c1f943a513de7a6c9bc9/server/src/main/java/org/elasticsearch/index/store/Store.java#L1323

**Pull request merged**